### PR TITLE
Fix error when the foreign field is overwritten by the relationship

### LINKED ISSUE #3: the foreign field is overwritten by the relationship
### Problem

Test case:
```typescript
test("foreign fields & relations", () => {
  const shape = getShape<Check>()({
    id: true,
    place_id: true,
    user_id: true,
    place: { _: "place_id", id: true, title: true },
    user: { _: "user_id", id: true, name: true },
  });

  const fields = getFields(shape);

  expect(fields).toBe("id,place_id,user_id,place:place_id(id,title),user:user_id(id,name)");
});
```

Error result

```typescript
Expected: "id,place_id,user_id,place:place_id(id,title),user:user_id(id,name)"
Received: "id,place:place_id_id,user:user_id_id,place:place_id(id,title),user:user_id(id,name)"
```

### PR DESCRIPTION
**Pull request by jedymatt**
- Added test case
- Fix error that overwrites the foreign field

### CODE REVIEW

**Comment by jedymatt** [comment]
### Problem Before The Fix

Test case:
```typescript
test("foreign fields & relations", () => {
  const shape = getShape<Check>()({
    id: true,
    place_id: true,
    user_id: true,
    place: { _: "place_id", id: true, title: true },
    user: { _: "user_id", id: true, name: true },
  });

  const fields = getFields(shape);

  expect(fields).toBe("id,place_id,user_id,place:place_id(id,title),user:user_id(id,name)");
});
```


Error result

```typescript
Expected: "id,place_id,user_id,place:place_id(id,title),user:user_id(id,name)"
Received: "id,place:place_id_id,user:user_id_id,place:place_id(id,title),user:user_id(id,name)"
```

**Comment by mzalevski** [comment]
I will look into it. Thanks.

File: index.test.ts [hunk 1]
```diff
@@ -166,3 +166,17 @@ test("all fields, relations & omissions", () => {
...
+
+test("foreign fields & relations", () => {
+  const shape = getShape<Check>()({
+    id: true,
+    place_id: true,
+    user_id: true,
+    place: { _: "place_id", id: true, title: true },
+    user: { _: "user_id", id: true, name: true },
+  });
+
```
Comment (mzalevski) [inline index.test.ts:178]: you should also add a type test:
```typescript
  expectType<{
    id: string;
    place_id: string;
    user_id: string;
    place: { id: string; title: string };
    user: { id: string; name: string };
  }>(shape);
  ```

File: src/index.ts [hunk 2]
```diff
@@ -77,15 +77,29 @@ export const getFields = (shape: Record<string, any>) => {
     return undefined;
   })
     .replace(/[": \n]/g, "")
-    .replace(/\{/g, "(")
-    .replace(/\}/g, ")")
+    // .replace(/\{/g, "(")
+    // .replace(/\}/g, ")")
```
Comment (mzalevski) [inline src/index.ts]: Leaving such comments is unnecessary. We can always check the previous version in git.

File: src/index.ts [hunk 3]
```diff
@@ -77,15 +77,29 @@ export const getFields = (shape: Record<string, any>) => {
...
+    return fields.replace(
+      new RegExp(Object.keys(joinsModifiedKeys).join("|"), "g"),
+      (m) => {
+        let trimKey = m.slice(0, -1); // removed the last `{`
+        return `${trimKey}:${joinsModifiedKeys[m]}(`; // appended the `(`
+      }
+    )
+      .replace(/\{/g, '(')
+      .replace(/\}/g, ')');
+  }
```
Comment (mzalevski) [inline src/index.ts]: This solution works, but there is a better one.

code to change on master:
```typescript
  if (Object.keys(joins).length > 0)
    return fields.replace(
      new RegExp(Object.keys(joins).join("|"), "g"),
      (m) => `${m}:${joins[m]}`
    );
```
another solution to the problem:
```typescript
  if (Object.keys(joins).length > 0)
    return fields.replace(
      new RegExp(Object.keys(joins).map((j) => `${j}(?=\\()`).join("|"), "g"),
      (m) => `${m}:${joins[m]}`
    );
```
as you can see, the only thing we need to fix the issue is a regex 'positive lookahead' on every join:
```
.map((j) => `${j}(?=\\()`)
```
the above means "match only if the next char is `(`"

the changes I'd like to see in this pull request look like [this](https://github.com/mzalevski/postgrest-js-tools/compare/master...the-foreign-field-is-overwritten-by-the-relationship)

**Review by mzalevski — CHANGES_REQUESTED** [review]

**Review by mzalevski — APPROVED** [review]

**Comment by mzalevski** [comment]
oh I forgot that the release is not triggered if there is no conventional commit prefix like `fix` in this case. I will do some magic to make up for it. Anyway, thanks for help @jedymatt :)

**Comment by jedymatt** [comment]
@mzalevski Thank you too for creating this awesome library as it can speed up my task. ヾ(≧▽≦*)o

**Pull request merged**